docs: ui.action_menu (#928)
deephaven/deephaven-plugins: plugins/ui/docs/components/action_menu.md

# Action Menu

An action menu merges an action button with a menu for easy access to more actions.

## Example

```python
from deephaven import ui

my_action_menu_basic = ui.action_menu("Cut", "Copy", "Paste")
```

## UI recommendations

> Consider using an [`action_group`] to group multiple action buttons together. 

Reviewer: no link

    selected_action, set_selected_action = ui.use_state("")
    return [
        ui.action_menu(
            "Cut",
            "Copy",
            "Paste",
            on_action=set_selected_action,
        ),
        ui.text(f"The action you have selected is: {selected_action}"),
    ]

my_action_menu_on_change_example = ui_action_menu_on_change_example()
```

## Sections

The action menu supports sections that group options. Sections can be used by wrapping groups of items in a `ui.section` element. Each section takes a `title` and `key` prop.

```python
from deephaven import ui

my_action_menu_section_example = ui.action_menu(
    ui.section(ui.item("Write"), ui.item("Append"), title="Addition"),
    ui.section(ui.item("Erase"), ui.item("Remove"), title="Deletion"),
)
```

## Complex items

Items within an action menu include additional content to better convey options. You can add icons, avatars, and descriptions to the children of an `ui.item`. When adding a description, set the `slot` prop to "description" to differentiate between the text elements.

```python
from deephaven import ui

my_action_menu_complex_items_example = ui.action_menu(
    ui.item(

>         ui.icon("vsGithubAlt"),
>         ui.text("Github"),
>         ui.text("Github Option", slot="description"),
>         text_value="Github",
>     ),
>     ui.item(
>         ui.icon("vsAzureDevops"),
>         ui.text("Azure"),
>         ui.text("Azure Option", slot="description"),
>         text_value="Azure",

Reviewer 2: ```suggestion
        ui.icon("github_alt"),
        ui.text("Github"),
        ui.text("Github Option", slot="description"),
        text_value="Github",
    ),
    ui.item(
        ui.icon("azure_devops"),
        ui.text("Azure"),
        ui.text("Azure Option", slot="description"),
        text_value="Azure",
```

Icons now support python snake_case. For backwards compatibility they also support JS style icon names, but we should show python snake case style icon names from now on.

```python
from deephaven import ui

my_action_menu_basic = ui.action_menu("Cut", "Copy", "Paste", is_quiet=True)
```

## Disabled State

An action menu can be disabled to prevent user interaction through the `is_disabled`. This is useful when the an action menu is not currently available, but the button should still be visible.

```python
from deephaven import ui

my_action_menu_basic = ui.action_menu("Cut", "Copy", "Paste", is_disabled=True)
```

## Flip

By default, the ActionMenu automatically flips direction when space is limited. To disable this behavior, set the `should_flip` prop to false

```python
from deephaven import ui

@ui.component
def ui_action_menu_flip_examples():
    return [
        ui.action_menu(
            "Cut", "Copy", "Paste", align="start", direction="top", should_flip=True
        ),
        ui.action_menu(
            "Cut", "Copy", "Paste", align="start", direction="top", should_flip=False
        ),
    ]

my_action_menu_flip_examples = ui_action_menu_flip_examples()
```

## Align and direction

> The `align` prop positions the action menu relative to the trigger, while the `direction` prop determines the direction in which the action menu will render.

Reviewer: "relative to the trigger" - I'm not 100% sure what that means. Is the usual trigger simply clicking the button? or do you mean the action triggered

Author: I mean relative to the button, but don't want to add confusion by mentioning a "button", since the children are items, not buttons. So the best way I thought to get around that is to generalize it by stating the "trigger".

Reviewer: I think button is fine, but defer to Don and the other reviewers.

Author: @dsmmcken Any thoughts?

Reviewer 2: parent button? trigger was fine with me though.

Author: Yeah, either relative to the "trigger", or relative to the "button", I personally think trigger makes sense?

```python
from deephaven import ui

@ui.component
def ui_action_menu_align_direction_examples():
    return [
        ui.action_menu("Cut", "Copy", "Paste", align="start"),
        ui.action_menu(
            "Cut", "Copy", "Paste", align="start", direction="top", should_flip=False
        ),
        ui.action_menu(
            "Cut",
            "Copy",
            "Paste",
            align="start",
            direction="start",
        ),
        ui.action_menu(
            "Cut",
            "Copy",
            "Paste",
            align="end",
            direction="end",
        ),
    ]

my_action_menu_align_direction_examples = ui_action_menu_align_direction_examples()
```

## Open

> The `is_open` and `default_open` props on the action menu control whether the menu is open by default. They apply controlled and uncontrolled behavior on the Menu respectively.

Reviewer: ```suggestion
The `is_open` and `default_open` props on the action menu control whether the menu is open by default. They apply controlled and uncontrolled behavior on the menu, respectively.
```

```python
from deephaven import ui

> @ui.component
> def ui_action_menu_open_example():
>     is_open, set_is_open = ui.use_state("")
>     return ui.action_menu(
>         "Cut",
>         "Copy",
>         "Paste",
>         on_open_change=set_is_open,
>     )

Reviewer 2: ```suggestion
@ui.component
def ui_action_menu_open_example():
    is_open, set_is_open = ui.use_state(True)
    return ui.action_menu(
        "Cut",
        "Copy",
        "Paste",
        is_open=is_open,
        on_open_change=set_is_open,
    )
```

I would actually show an example with the props you listed.

    ]

my_action_menu_flip_examples = ui_action_menu_flip_examples()
```

## Align and direction

The `align` prop positions the action menu relative to the trigger, while the `direction` prop determines the direction in which the action menu will render.

```python
from deephaven import ui

@ui.component
def ui_action_menu_align_direction_examples():
    return [
        ui.action_menu("Cut", "Copy", "Paste", align="start"),
        ui.action_menu(
            "Cut", "Copy", "Paste", align="start", direction="top", should_flip=False
        ),
        ui.action_menu(
            "Cut",
            "Copy",
            "Paste",
            align="start",
            direction="start",
        ),
        ui.action_menu(
            "Cut",
            "Copy",
            "Paste",
            align="end",
            direction="end",
        ),
    ]

my_action_menu_align_direction_examples = ui_action_menu_align_direction_examples()
```

## Open

The `is_open` and `default_open` props on the action menu control whether the menu is open by default. They apply controlled and uncontrolled behavior on the menu respectively.

```python
from deephaven import ui

@ui.component

> def ui_action_menu_open_examples():
>     is_open, set_is_open = ui.use_state(False)
>     return [
>         ui.action_menu(

Reviewer 2: ```suggestion
def ui_action_menu_open_examples():
    is_open, set_is_open = ui.use_state(False)
    return [
        ui.text(f"Controlled menu open state: {is_open}"),
        ui.action_menu(
```

We should try to visually show changes if we can, in this case we are trying to show that you can use a controlled component, so show the state being changed.

    ui.section(ui.item("Write"), ui.item("Append"), title="Addition"),
    ui.section(ui.item("Erase"), ui.item("Remove"), title="Deletion"),
)
```

## Complex items

Items within an action menu include additional content to better convey options. You can add icons, avatars, and descriptions to the children of an `ui.item`. When adding a description, set the `slot` prop to "description" to differentiate between the text elements.

```python
from deephaven import ui

my_action_menu_complex_items_example = ui.action_menu(
    ui.item(
        ui.icon("github_alt"),
        ui.text("Github"),
        ui.text("Github Option", slot="description"),
        text_value="Github",
    ),
    ui.item(
        ui.icon("azure_devops"),
        ui.text("Azure"),
        ui.text("Azure Option", slot="description"),
        text_value="Azure",
    ),
)
```

## Quiet State

The `is_quiet` prop makes an action menu "quiet". This can be useful when the action menu and its corresponding styling should not distract users from surrounding content.

```python
from deephaven import ui

my_action_menu_basic = ui.action_menu("Cut", "Copy", "Paste", is_quiet=True)
```

## Disabled State

> An action menu can be disabled to prevent user interaction through the `is_disabled` prop. This is useful when the action menu is not currently available, but the button should still be visible.

Reviewer: ```suggestion
Through the ' is_disabled ' prop, an action menu can be disabled to prevent user interaction. This is useful when the action menu is currently unavailable, but the button should still be visible.
```
